src/com/ubcsolar/sim/SimController.java

```diff
@@ -115,6 +115,110 @@ public void runSimulation(int laps, long startTime)
 
 		this.mySession.sendNotification(new NewSimulationReportNotification(toSend));
 	}
+	
+	public void runSimulationWithManualSpeeds(int laps, long startTime, Map<GeoCoord, Map<Integer, Double>> manualSpeedProfile)
+			throws NoForecastReportException, NoLoadedRouteException, NoLocationReportedException,
+			NoCarStatusException {
+		if (laps <= 0) {
+			throw new IllegalArgumentException("Number of Laps too low, must go at least 1 lap");
+		}
+		// Compile all the information we need.
+		LocationReport lastReported = this.mySession.getMapController().getLastReportedLocation();
+		if (lastReported == null) {
+			throw new NoLocationReportedException();
+		}
+		Route routeToTraverse = this.mySession.getMapController().getAllPoints();
+
+		if (routeToTraverse == null) {
+			throw new NoLoadedRouteException();
+		}
+		TelemDataPacket lastCarReported = this.mySession.getMyCarController().getLastTelemDataPacket();
+		if (lastCarReported == null) {
+			throw new NoCarStatusException();
+		}
+
+		SimResult result = null;
+		List<SimFrame> resultFrames = new ArrayList<SimFrame>(); //list of sim frames
+		//idea is to use runSimV2 for every part of the route where the car travels at constant speed until the speed change
+		//the point with the different speed would be the last point, and the point after would be the start of the next simulation
+		GeoCoord start = routeToTraverse.getTrailMarkers().get(1); //first point of the simulation (need to ask why its 1, running point 0 to 1 is always a fail)
+		GeoCoord end; //last point of the simulation
+		TreeMap<Integer, ForecastIO> inflectionPoints = mySession.getMyWeatherController()
+				.findInflectionPoints(routeToTraverse, currentForecastReport.getForecasts());
+		ForecastIO current_forecast = inflectionPoints.get(inflectionPoints.keySet().toArray()[0]);
+		long nextStartTime = startTime;
+		long totalTime = 0;
+		
+		for (int i = 1; i <= laps; i++) {
+			//iterate through all the points except for the last point (perserve last point for one last runSimV2 call since car may be traveling at constant speeds
+			//until the end
+			for (int point_index = 1; point_index < routeToTraverse.getTrailMarkers().size(); point_index++) {
+				//check to see if an inflection point has reached and a new forecast is used
+				if (inflectionPoints.keySet().contains(i)) {
+					current_forecast = inflectionPoints.get(i);
+				}
+				
+				//check to see if there is a speed change
+				//if there is, run a simulation and update list of sim frames, new start speed, new start point, start time and car status using results
+				if (Math.abs(manualSpeedProfile.get(start).get(i) - manualSpeedProfile.get(routeToTraverse.getTrailMarkers().get(point_index)).get(i)) > 0.000001 ||
+						point_index == routeToTraverse.getTrailMarkers().size()) {
+					end = routeToTraverse.getTrailMarkers().get(point_index);
```

Comment:
does this work if we hit the `point_index == routeToTraverse.getTrailMarkers().size()` condition in the if statement? wouldn't that be out of bounds?